rfcs/rfc0010.md

```diff
@@ -0,0 +1,157 @@
+# A built-in for getting key-and-value pairs
+
+## Preamble
+
+    Author:  Ricardo Signes <rjbs@semiotic.systems>
+    Sponsor:
+    ID:      0010
+    Status:  Draft
+
+## Abstract
+
+This RFC proposes `kv`, a new function for returning the keys and values of a
+container.  Among other things, this allows key/value iteration on arrays as
+easily on hashes.
+
+## Motivation
+
+Right now, the `keys` and `values` functions operate on both arrays and hashes,
+giving the same semantics, if we accept that array indexes are effectively the
+keys of the array.  Pair slices, similarly, produce key/value pairs of array
+entries.  The sole missing component is that the behavior of a hash in list
+context is not easily reproduced.
+
+Providing this doesn't just fulfill a sense of completionism, but also has an
+immediate practical benefit.  With v5.36.0 poised to add n-at-a-time foreach,
+easily getting a list of index/value pairs from an array makes iteration over
+them easy.
+
+## Rationale
+
+If we start with the specific case of iterating over the indexes and values of
+an array using two-target foreach, we might write this:
+
+```perl
+for my ($i, $value) (%array[ keys @array ]) {
+  say "$i == $value";
+}
+```
+
+This is tolerable, but a bit verbose.  If we bury our target array deep in a
+structure, we get this:
+
+```perl
+for my ($i, $value) ($alpha->{beta}->[0]->%[ keys $alpha->{beta}->[0]->@* ]) {
+  say "$i == $value";
+}
+```
+
+This is pretty bad.
+
+With `kv`, we write this:
+
+```perl
+for my ($i, $value) (kv $alpha->{beta}->[0]->@*) {
+  say "$i == $value";
+}
+```
+
+This is probably about as simple as this can get without some significant new
+addition to the language.
+
+## Specification
+
+Like `keys` or `values`, `kv` can be called with a hash expression or array
+expression.  Other types of arguments are compile time errors.
+
+In scalar context, `kv` returns the number of entries in its argument, just
+like `keys` or `values`.
+
+In list context, `kv %h` is identical in behavior to `%h` in list context.
+
+In list context, `kv @arr` returns a list twice the size of the array, zipping
+together the keys and values.  To match existing behaviors, the keys are
+mutable copies and the values are aliases.  This requires the following work:
+
+```
+  DB<1> @arr = (1..10)
+
+  DB<2> x \@arr
+0  ARRAY(0x7f9784b09ed8)
+   0  1
+   1  2
+   2  3
+   3  4
+   4  5
+   5  6
+   6  7
+   7  8
+   8  9
+   9  10
+
+  DB<3> $_ ++ for kv @arr
+
+  DB<4> x \@arr
+0  ARRAY(0x7f9784b09ed8)
+   0  2
+   1  3
+   2  4
+   3  5
+   4  6
+   5  7
+   6  8
+   7  9
+   8  10
+   9  11
+```
+
+## Backwards Compatibility
+
+There should be no significant backwards compatibility concerns.  `kv` will be
+imported only when requested.  Static analysis tools may need to be updated.
+
+For polyfill in older perls, I don't believe a pure-perl implementation of `kv`
+can be provided while maintaining its aliasing behavior.  This may not be true.
```

Comment:
Not a pureperl one, but an XS one can do so easily.

Reply:
List lvalue functions are pretty questionable, but they do make this possible:

```
sub kv (\[%@]) :lvalue {
  Scalar::Util::reftype($_[0]) eq 'ARRAY'
    ? (wantarray ? $_[0]->%[keys $_[0]->@*] : $_[0]->@*)
    : $_[0]->%*;
}
```